### PR TITLE
Add conditions to notices

### PR DESCRIPTION
**Pull request by ThoWagen**
This PR adds the option to add conditions to notices.

It was mainly taken from here https://github.com/MSU-Libraries/catalog/blob/daa34afd1d1ab8ff530fb352af8deecf6a2fa9d8/vufind/module/Catalog/src/Catalog/View/Helper/Root/Notices.php#L196-L324. That is why I noted MSU in the copyright and @natecollins as author. But the condition types are now pluggable as the "condition handlers".

To keep this PR small, it only provides a handler for dates and one for static strings. The latter is used for testing and I don't know if there is actually a real world example where it would make sense to use it. @natecollins @meganschanz please let me know, if there is and I will remove the doc comment that says it is only used for tests.

More handlers will be added in a followup PR.

I want to use the conditions in upcoming PRs for other features and that's why I believe that it would be better to document the conditions in a wiki page instead of the Notices.yaml. @demiankatz will you create a page for that, so that I can insert the documentation? Not sure, if and how I can do that myself.

- [x] Add wiki page for conditions and update comment in Notices.yaml

### CODE REVIEW

**Comment by natecollins** [comment]
In our [sample BannerNotices config file](https://github.com/MSU-Libraries/catalog/blob/main/vufind/local/config/vufind/BannerNotices.yaml) we have documentation which includes a few examples of using the conditional logic.

For a real world use of the `string` type, one example in our sample config is paired with a `filetype` check value to display a message conditionally if that [string exists as a file/symlink](https://github.com/MSU-Libraries/catalog/blob/main/vufind/local/config/vufind/BannerNotices.yaml#L136-L141).

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * @category VuFind
+ * @package  Condition_Handlers
+ * @author   Nathan Collins <colli372@msu.edu>
+ * @author   Thomas Wagener <wagener@hebis.uni-frankfurt.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFind\Condition\Handlers;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: I would change this namespace to VuFind\Condition\Handler -- we generally use singular names in namespaces.

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            $condition['comparator'] ?? '',
+            $baseValue,
+            $checkedValues
+        );
+    }
+
+    /**
+     * Get base value to check.
+     *
+     * @param array $condition Optional used for handler specific parameters
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: This says "optional" but appears to be mandatory.

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+     * @throws ConditionException
+     */
+    abstract protected function getBaseValue(array $condition): string;
+
+    /**
+     * Evaluate a single set of comparison conditions.
+     *
+     * @param string   $comparator    The type of comparison to evaluate
+     * @param string   $baseValue     The value to validate
+     * @param string[] $checkedValues Values for the comparison which will result in success
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: This function comment is a bit confusing; maybe better to align with the description in the array key documentation above.

File: module/VuFind/src/VuFind/Condition/Handlers/StringHandler.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+     *
+     * @param array $condition Optional used for handler specific parameters
+     *
+     * @return string
+     * @throws ConditionException
+     */
+    protected function getBaseValue(array $condition): string
+    {
+        $string = $condition['string'] ?? null;
+        if ($string === null) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handlers/StringHandler.php]: Is it also worth doing a type check here, in case something weird is passed through the array?

File: module/VuFind/src/VuFind/Condition/Handler/Date.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Get base value to check.
+     *
+     * @param array $condition Optional used for handler specific parameters
+     *
+     * @return string
+     */
+    protected function getBaseValue(array $condition): string
+    {
+        return (new \DateTime())->format('Y-m-d');
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handler/Date.php:57]: Should we allow a DateTime to be passed through the condition for non-default comparisons?

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+abstract class AbstractBase implements ConditionHandlerInterface
+{
+    /**
+     * Check if a condition is met.
+     *
+     * Conditions are represented as an associative array with the following required keys:
+     * - type: identifier of the condition handler
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: Should we note that `type` is used by the condition manager to route requests but can be ignored by the actual handlers? I can't think of a situation where the handler itself would look at the type, unless you think we might have a multi-purpose handler that would get registered under multiple aliases.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen and @natecollins! I haven't fully reviewed this, but I left a few comments below based on some initial reading.

I've also created a wiki page for you: https://vufind.org/wiki/development:plugins:condition_handlers

I generally favor documentation in config files rather than the wiki so that the project is self-contained and so that version-specific documentation stays bundled with the associated code... but I see why in this case it wouldn't make sense to document the condition handlers in the Notices.yaml since they could be used in other contexts as well... so I think using the wiki is okay, as long as we clearly indicate which versions various handlers were added in.

Please let me know if you need anything else from me; I'll review more thoroughly after you've had a chance to respond to this initial review.

**Comment by ThoWagen** [comment]
> In our [sample BannerNotices config file](https://github.com/MSU-Libraries/catalog/blob/main/vufind/local/config/vufind/BannerNotices.yaml) we have documentation which includes a few examples of using the conditional logic.
> 
> For a real world use of the `string` type, one example in our sample config is paired with a `filetype` check value to display a message conditionally if that [string exists as a file/symlink](https://github.com/MSU-Libraries/catalog/blob/main/vufind/local/config/vufind/BannerNotices.yaml#L136-L141).

Wouldn't it make more sense to add filetype conditions as condition types instead of comparison?

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php (hunk 1 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: Makes sense. I changed it

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php (hunk 2 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: It will always be passed, but does not need to be used. E.g. the Date handler does ignore it. I added the the suppression of the warning.

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php (hunk 3 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: I updated it. Is it better now?

File: module/VuFind/src/VuFind/Condition/Handlers/StringHandler.php (hunk 4 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handlers/StringHandler.php]: It certainly does not hurt. I added it.

File: module/VuFind/src/VuFind/Condition/Handler/Date.php (hunk 5 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handler/Date.php:57]: I'm not sure what you mean.

File: module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php (hunk 6 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handlers/AbstractBase.php]: It could be used to identify the type in some abstract handler method, but there are probably better ways to do whatever that might be for. So I added the hint.

**Comment by ThoWagen** [comment]
> I generally favor documentation in config files rather than the wiki so that the project is self-contained and so that version-specific documentation stays bundled with the associated code... but I see why in this case it wouldn't make sense to document the condition handlers in the Notices.yaml since they could be used in other contexts as well... so I think using the wiki is okay, as long as we clearly indicate which versions various handlers were added in.

We could add the documentation in the Notices.yaml instead and always refer to it in later use cases, if you prefer that.

File: module/VuFind/src/VuFind/Condition/Handler/Date.php (hunk 5 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/Condition/Handler/Date.php:57]: Right now, the base value is always the current DateTime, but might there be a situation where we would want to override this with some other arbitrary date and time, so `$condition` would include a date value that could be used instead of creating a new object?

It's entirely possible there is no reasonable use case that would require this -- I am just brainstorming. Since the StringHandler accepts an arbitrary string, I thought maybe the Date handler should be able to accept an arbitrary date.

**Comment by demiankatz** [comment]
> We could add the documentation in the Notices.yaml instead and always refer to it in later use cases, if you prefer that.

I think it's okay to use the wiki for now, but I am fine with whichever of the two solutions you feel most comfortable with.

**Comment by natecollins** [comment]
> Wouldn't it make more sense to add filetype conditions as condition types instead of comparison?

If that's your preference, I have no objection.

File: module/VuFind/src/VuFind/Condition/Handler/Date.php (hunk 5 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Condition/Handler/Date.php:57]: For the planned use cases (conditions coming from the configurations or the database), the values that are checked against are hard coded and having hard coded base values does not really make sense, because the conditions are either always true or always false. And if you want that for some reason, you can already use the string handler.

**Comment by ThoWagen** [comment]
@demiankatz I added the documentation to https://vufind.org/wiki/development:plugins:condition_handlers. But I agree that it would be much better if this would be self-contained in the project. Would it make sense to add it in Markdown format under a new directory "documentation" instead?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+class DateTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * Test true condition.
+     *
+     * @return void
+     */
+    public function testTrueMatching(): void
+    {
+        $today = (new \DateTime())->format('Y-m-d');
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php]: It's unlikely but not impossible for this test to fail if run too close to midnight. A couple of possible approaches:

1.) Try it in a loop and allow it to fail a maximum of one time; that should cover the edge case.

2.) Allow injection of a specific DateTime object as previously discussed (which is not to say I'm still arguing that we should -- just that if we did, it could have some advantages for controlling tests... however, the first option is likely easier).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen! See below for one minor concern about a test. Additionally, a couple of other thoughts/replies:

1.) Regarding your suggestion of including markdown documentation in the project itself, I think that could be a promising idea -- but I wouldn't want to do it just for this one use case. Let's keep the documentation in the wiki for now, but I think we should keep this idea in mind as the Project Management Committee moves forward with documentation improvements and updates. Maybe it's also worth a future Community Call discussion.

2.) Now that @natecollins has replied to your earlier messages, is there any further work you want to complete before we wrap up this PR?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php (hunk 7 above)
Comment (ThoWagen) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php]: Since the plan is to also add handlers for DateTime and Time, where that problem would be even bigger, I just changed this to check if the date is in a range of 2 days.

**Comment by ThoWagen** [comment]
> 1.) Regarding your suggestion of including markdown documentation in the project itself, I think that could be a promising idea -- but I wouldn't want to do it just for this one use case. Let's keep the documentation in the wiki for now, but I think we should keep this idea in mind as the Project Management Committee moves forward with documentation improvements and updates. Maybe it's also worth a future Community Call discussion.

Alright, sounds good to me.

> 2.) Now that @natecollins has replied to your earlier messages, is there any further work you want to complete before we wrap up this PR?

No, I will add filet type conditions as a handler in a followup PR.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php (hunk 7 above)
Comment (demiankatz) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Condition/Handler/DateTest.php]: Great idea -- and then you test some extra comparators in the process. :-)

**Review by demiankatz — APPROVED** [review]
Looks good to me now, @ThoWagen!
